### ISSUE TITLE
class "jxlsave" not found

### ISSUE THREAD

**Issue by euzada** [libvips/pyvips#467]
Hi @jcupitt,

I tried to save an image as jpeg XL but getting the following error:

```
pyvips.error.Error: no such operation jxlsave
  VipsOperation: class "jxlsave" not found
```

I compiled libvips for windows with ``` ./build.sh --with-jpegli ``` but didn't work. I saw that since version [v8.11.0](https://github.com/libvips/build-win64-mxe/releases/tag/v8.11.0) that ``` Add libjxl to 'all' group dependencies for JPEG XL support (https://github.com/libvips/build-win64-mxe/pull/28).```. I used 'all' version and still having the same issue. I cannot use jxlsave in pyvips.

pyvips version = 2.2.3
libvips version = vips-8.15.2

Any suggestion?

**Comment by jcupitt** [comment]
Hello @euzada,

Did you try the official win binary? It should include `jxlsave`.

https://github.com/libvips/build-win64-mxe/releases/tag/v8.15.2

You need 8.15.2 w64 all.

**Comment by euzada** [comment]
> Hello @euzada,
> 
> Did you try the official win binary? It should include `jxlsave`.
> 
> https://github.com/libvips/build-win64-mxe/releases/tag/v8.15.2
> 
> You need 8.15.2 w64 all.

Yes I did. Now it run without the error and generate the jxl file but the file is broken. The file size is large, tried to shrink the image but no success open it. Now I am facing another issue with libtiff that cannot test my script.

Will do another test later after I fix the tif_jpeg.c issues I am having.

**Comment by jcupitt** [comment]
Oh, strange, let me try here.

**Comment by kleisauke** [comment]
> jxlsave: error JxlEncoderProcessOutput

I could not reproduce this with the `vips-dev-w64-all-8.15.2.zip` binaries on my Windows 11 PC.

```console
$ python
Python 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> ref = pyvips.Image.new_from_file('zebra.jpg')
>>> ref.jxlsave('x.jxl', lossless=True)
>>> x = pyvips.Image.new_from_file('x.jxl', access='sequential')
>>> (ref - x).abs().max()
0.0
>>>
```

**Comment by euzada** [comment]
Hi,

I figure it out. I am having this image only for output image larger than 10kx10k image. For smaller images, it works, for larger it said jxlsave not found.

**Comment by jcupitt** [comment]
Could it be running out of memory? Until very recently, libjxl could use HUGE amounts of ram for larger images.

vipsdisp 3.0.4 for win is built against a recent libjxl. It could be worth checking memuse for that program when saving a large JXL image.

**Comment by kleisauke** [comment]
Indeed, it sounds like an OOM error. I was able to create a 100000x100000 image without any issues.
```console
$ python
Python 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.black(100000, 100000, bands=3)
>>> x.jxlsave('x.jxl')
>>> pyvips.Image.new_from_file('x.jxl', access='sequential')
<pyvips.Image 100000x100000 uchar, 3 bands, srgb>
>>>
```

**Comment by jcupitt** [comment]
I tried making a JXL image with libjxl 0.7:

```
$ /usr/bin/time -f %M:%e vips replicate nina.jpg x.jxl 2 3 
10983132:22.56
```

That's 11gb of memory for a 12k x 12k image. With 0.10.1 I see:

```
$ /usr/bin/time -f %M:%e vips replicate nina.jpg x.jxl 2 3 
1314856:70.23
```

A more reasonable 1.3gb, though it's noticeably slower for some reason.
